internal/controller/controllers/agentclusterinstall_controller.go

```diff
@@ -0,0 +1,112 @@
+)
+
+// AgentClusterInstallReconciler reconciles a AgentClusterInstall object
+type AgentClusterInstallReconciler struct {
+	client.Client
+	Log              logrus.FieldLogger
+	CRDEventsHandler CRDEventsHandler
+}
+
+// +kubebuilder:rbac:groups=extensions.hive.openshift.io,resources=agentclusterinstalls,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=extensions.hive.openshift.io,resources=agentclusterinstalls/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=extensions.hive.openshift.io,resources=agentclusterinstalls/finalizers,verbs=update
+
+func (r *AgentClusterInstallReconciler) Reconcile(origCtx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	ctx := addRequestIdIfNeeded(origCtx)
+	log := logutil.FromContext(ctx, r.Log).WithFields(
+		logrus.Fields{
+			"agent_cluster_install":           req.Name,
+			"agent_cluster_install_namespace": req.Namespace,
+		})
+
+	defer func() {
+		log.Info("AgentClusterInstall Reconcile ended")
+	}()
+
+	log.Info("AgentClusterInstall Reconcile started")
+
+	// Retrieve AgentClusterInstall
+	clusterInstall := &hiveext.AgentClusterInstall{}
+	if err := r.Get(ctx, req.NamespacedName, clusterInstall); err != nil {
+		log.WithError(err).Errorf("Failed to get resource %s", req.NamespacedName)
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// Retrieve ClusterDeployment
+	clusterDeploymentKubeKey := types.NamespacedName{
+		Namespace: clusterInstall.Namespace,
+		Name:      clusterInstall.Spec.ClusterDeploymentRef.Name,
+	}
+	clusterDeployment := &hivev1.ClusterDeployment{}
+	if err := r.Get(ctx, clusterDeploymentKubeKey, clusterDeployment); err != nil {
+		var errMsg string
+		if k8serrors.IsNotFound(err) {
+			apiErr := errors.New(fmt.Sprintf(
+				"ClusterDeployment with name '%s' in namespace '%s' not found. Ensure that the CRD is already applied.",
+				clusterDeploymentKubeKey.Name, clusterDeploymentKubeKey.Namespace))
+			err = common.NewApiError(http.StatusBadRequest, apiErr)
+			log.WithError(err).Error(errMsg)
+		} else {
+			errMsg = fmt.Sprintf("failed to get ClusterDeployment with name '%s' in namespace '%s'",
+				clusterDeploymentKubeKey.Name, clusterDeploymentKubeKey.Namespace)
+			log.WithError(err).Error(errMsg)
+		}
+
+		// Set conditions
+		clusterSpecSynced(clusterInstall, err)
+		setClusterConditionsUnknown(clusterInstall)
+
+		if updateErr := r.Status().Update(ctx, clusterInstall); updateErr != nil {
+			log.WithError(updateErr).Error("failed to update AgentClusterInstall Status")
+			return ctrl.Result{Requeue: true}, nil
+		}
+		if err != nil && !IsUserError(err) {
+			return ctrl.Result{RequeueAfter: defaultRequeueAfterOnError}, nil
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *AgentClusterInstallReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&hiveext.AgentClusterInstall{}).
+		Watches(&source.Channel{Source: r.CRDEventsHandler.GetAgentUpdates()},
```

Comment:
Not sure that this watch is required. 
This is usually notifications from the backend, not something that should affect the logic of this controller.
Unless i'm missing something?

Reply:
I agree, watch on AgentClusterInstall for this is enough.

Reply:
Right, removed.